src/main/java/com/chargerrobotics/utils/ArduinoSerialReceiver.java

```diff
@@ -0,0 +1,70 @@
+package com.chargerrobotics.utils;
+
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Map.Entry;
+
+import com.fazecast.jSerialComm.SerialPort;
+
+public class ArduinoSerialReceiver {
+
+	private static final HashMap<byte[], ArduinoListener> responseHeaders = new HashMap<byte[], ArduinoListener>();
+	private static final ArrayList<ArduinoSerial> serialPorts = new ArrayList<ArduinoSerial>();
+
+	public static void init() {
+		for (SerialPort availablePort : SerialPort.getCommPorts()) {
+			if (availablePort.toString().contains("USB-to-Serial")) {
+				serialPorts.add(new ArduinoSerial(availablePort.getSystemPortName()));
+			}
+		}
+	}
+
+	public static void poll() {
+		for (ArduinoSerial port : serialPorts) {
+			port.poll();
+		}
+	}
+
+	public static void registerListener(ArduinoListener listener, byte header1, byte header2) {
+		responseHeaders.put(new byte[] { header1, header2 }, listener);
+	}
+
+	protected static ArduinoListener getListener(byte[] header) {
+		for (Entry<byte[], ArduinoListener> entry : responseHeaders.entrySet()) {
+			byte[] test = entry.getKey();
+			if (test[0] == header[0] && test[1] == header[1])
+				return entry.getValue();
+		}
+		return null;
+	}
+
+	public static abstract class ArduinoListener {
+		private long lastReceived = 0;
+		private long expiry = 50;
+
+		public abstract void receiveData(ArduinoSerial serial, ByteBuffer data);
```

Comment:
This interface will expose the ArduinoSerial's internal ByteBuffer to every listener. This leads to the possibility of the entire port's data being corrupted unless every listener understands and respects whatever behavior you expect of them. It also requires the authors of every listener to understand how to extract all of their message data (and no more) from the ByteBuffer, and how to ensure that they don't crash when a corrupted or incomplete message is received.

A safer and more robust approach is to provide a message to listeners that does not allow them any access to the port's internal data structures, and that contains exactly the data that is part of that message.

Reply:
With the use of `asReadOnlyBuffer`, there is no more danger of a listener consuming past the end of its message, since the readonly buffer has a separate position pointer from the original buffer.